### PR TITLE
Fixes #7982 - Makes Users and Orgs API GET endpoint results standard to Github

### LINKED ISSUE #7982: missing API call get /admin/users/{username}
- Gitea version (or commit ref): 1.9.1
- Git version:
- Operating system: 
- Database (use `[x]`):
  - [ ] PostgreSQL
  - [ ] MySQL
  - [ ] MSSQL
  - [x] SQLite
- Can you reproduce the bug at https://try.gitea.io:
  - [ ] Yes (provide example URL)
  - [ ] No
  - [x] Not relevant
- Log gist:

## missing an API call to get all informations for an existing user

the call `GET /users/{username}` sends only a few user properties, 
but not enough to change some values (eg. allow_create_organization, or active) 
and send this changes back to server by API call `PATCH /admin/users/{username}`

Short: i missing the API call `GET /admin/users/{username}`

### PR DESCRIPTION
**Pull request by richmahn**
Fixes #7982 

This makes it so we are getting all the information about a user or org like one gets with GitHub. It makes it so public API endpoints work, but respects email showing unless one is admin.

Also fixes a security hole where, as you can see in the diff of models/org.go, it was showing an org even if it was not set to public (was only going by if the member was public or not, but needs to go by if the org is private or not)

Also makes it so the API convert.go methods for model.User -> api.user uses the model.User.APIFormat() (reduces redundant code) but makes sure email is only show if API has a user logged in.

### CODE REVIEW

File: models/user.go [hunk 1]
```diff
@@ -6,6 +6,7 @@
 package models
 
 import (
+	"code.gitea.io/gitea/modules/markup"
```
Comment (lafriks) [inline models/user.go]: wrong import order

File: routers/api/v1/convert/convert.go [hunk 2]
```diff
@@ -5,13 +5,13 @@
 package convert
 
 import (
+	"code.gitea.io/gitea/modules/setting"
```
Comment (lafriks) [inline routers/api/v1/convert/convert.go]: wrong import order

File: modules/structs/org.go [hunk 3]
```diff
@@ -4,16 +4,27 @@
 
 package structs
 
+import "time"
```
Comment (lafriks) [inline modules/structs/org.go]:
```suggestion
import (
	"time"
)
```

**Review by lafriks — CHANGES_REQUESTED** [review]

File: models/user.go (hunk 1 above)
Comment (richmahn) [reply models/user.go]: Thanks.

File: modules/structs/org.go (hunk 3 above)
Comment (richmahn) [reply modules/structs/org.go]: Going to stop letting my IDE do the imports....will turn that off. 😡

**Comment by lunny** [comment]
Is this a break change?

File: models/user.go [hunk 4]
```diff
@@ -224,16 +236,36 @@ func (u *User) GetEmail() string {
...
-		FullName:  u.FullName,
-		Email:     u.GetEmail(),
-		AvatarURL: u.AvatarLink(),
-		Language:  u.Language,
-		IsAdmin:   u.IsAdmin,
-		LastLogin: u.LastLoginUnix.AsTime(),
-		Created:   u.CreatedUnix.AsTime(),
+		ID:               u.ID,
+		UserName:         u.Name,
+		FullName:         markup.Sanitize(u.FullName),
```
Comment (guillep2k) [inline models/user.go]: Why is this necessary? Doesn't the JSON layer take care of this?

File: modules/structs/user.go [hunk 5]
```diff
@@ -20,16 +20,52 @@ type User struct {
...
 	// User locale
 	Language string `json:"language"`
 	// Is the user an administrator
-	IsAdmin bool `json:"is_admin"`
+	IsAdmin bool `json:"site_admin"`
 	// swagger:strfmt date-time
 	LastLogin time.Time `json:"last_login,omitempty"`
 	// swagger:strfmt date-time
-	Created time.Time `json:"created,omitempty"`
+	Created time.Time `json:"created"`
```
Comment (guillep2k) [inline modules/structs/user.go]: If the `created` field is empty (I don't know, perhaps an old version of Gitea or some migration), will this break now?

File: routers/api/v1/api.go [hunk 6]
```diff
@@ -513,12 +513,20 @@ func RegisterRoutes(m *macaron.Macaron) {
...
+					m.Get("/:target", user.CheckFollowing)
+				})
+				m.Get("/gpg_keys", user.ListGPGKeys)
 				m.Get("/heatmap", mustEnableUserHeatmap, user.GetUserHeatmapData)
-
+				m.Get("/keys", user.ListPublicKeys)
+				m.Get("/orgs", org.ListUserOrgs)
 				m.Get("/repos", user.ListUserRepos)
+				m.Get("/starred", user.GetStarredRepos)
+				m.Get("/subscriptions", user.GetWatchedRepos)
```
Comment (guillep2k) [inline routers/api/v1/api.go]: I wonder if all this info is public for all users?

File: routers/api/v1/api.go [hunk 7]
```diff
@@ -527,23 +535,6 @@ func RegisterRoutes(m *macaron.Macaron) {
 			})
 		})
```
Comment (guillep2k) [inline routers/api/v1/api.go:538]: `}, reqToken())` needed somewhere?

File: routers/api/v1/convert/convert.go [hunk 8]
```diff
@@ -205,15 +205,25 @@ func ToDeployKey(apiLink string, key *models.DeployKey) *api.DeployKey {
...
-		UserName:    org.Name,
-		FullName:    org.FullName,
-		Description: org.Description,
-		Website:     org.Website,
-		Location:    org.Location,
-		Visibility:  org.Visibility.String(),
+		ID:               org.ID,
+		AvatarURL:        org.AvatarLink(),
+		UserName:         org.Name,
+		FullName:         org.FullName,
```
Comment (guillep2k) [inline routers/api/v1/convert/convert.go]: `Sanitize()` was used in a previous change. It's not needed here?

**Review by guillep2k — COMMENTED** [review]
Just some questions. I'm not terribly familiar with the code or the API.

File: routers/api/v1/api.go (hunk 6 above)
Comment (richmahn) [reply routers/api/v1/api.go]: Yes, GitHub makes that public for all users too. See https://api.github.com/users/richmahn, https://api.github.com/users/richmahn/starred, https://api.github.com/users/richmahn/subscriptions, https://api.github.com/users/richmahn/gpg_keys, https://api.github.com/users/richmahn/keys, etc. These are public keys which a client may need to see/use.

File: models/user.go (hunk 4 above)
Comment (richmahn) [reply models/user.go]: I didn't write that. This was just moved from convert.go so that both models/user.go and api/v1/convert/covert.go use the same implementation.

File: routers/api/v1/api.go (hunk 6 above)
Comment (richmahn) [reply routers/api/v1/api.go]: I even searched to see why these were required the user to be authenticated, yet never found a reason. I assume it was just done that way because Gogs originally was more for just a single or few users.

File: models/user.go (hunk 4 above)
Comment (richmahn) [reply models/user.go]: Hmm, I think you're right though. It was copied from convert.go since I first did my changes there and then moved that whole structure population into here. Will remove.

File: routers/api/v1/api.go (hunk 7 above)
Comment (richmahn) [reply routers/api/v1/api.go:538]: Done some extensive testing on all http://localhost:3000/api/v1/users/* API calls and all seems to be secure, in that everything public on GitHub is also the same in my branch of Gitea. I will keep testing this though, just to make sure all is secure.

File: routers/api/v1/convert/convert.go (hunk 8 above)
Comment (richmahn) [reply routers/api/v1/convert/convert.go]: Yeah, I never made that change, as I wrote in another comment. I agree with you that it is handled in the JSON serialization.

**Comment by richmahn** [comment]
@lunny Shouldn't be a break change, but it does make all the users/* API calls public like GitHub has them. Doesn't make sense to just have to be a Gitea user to query info about a user. But that is very open for discussion if we want to be more private than GitHub. As I posted in a review thread above, you can see these links for my account on GitHub without a token or authentication:

https://api.github.com/users/richmahn
https://api.github.com/users/richmahn/starred https://api.github.com/users/richmahn/subscriptions https://api.github.com/users/richmahn/gpg_keys
https://api.github.com/users/richmahn/keys

Also, I added data to the GET responses for Users and Orgs, nothing removed.

File: routers/api/v1/api.go (hunk 7 above)
Comment (guillep2k) [reply routers/api/v1/api.go:538]: Oh, I see... the whole point was to allow annonymous access. 🤦‍♂

Comment (richmahn) [reply routers/api/v1/api.go:538]: Actually though your comment did catch something, that the tokens API was no longer usable with a token. I have fixed that.

File: integrations/api_user_orgs_test.go [hunk 9]
```diff
@@ -5,6 +5,7 @@
 package integrations
 
 import (
+	"code.gitea.io/gitea/modules/setting"
```
Comment (lunny) [inline integrations/api_user_orgs_test.go]: wrong import sequence.

Comment (richmahn) [reply integrations/api_user_orgs_test.go]: Thanks. Fixed.

**Comment by guillep2k** [comment]
[8086.zip](https://github.com/go-gitea/gitea/files/3695398/8086.zip)

Hi, @richmahn. Try this fix. The problem was that calling functions like `user.APIFormat()` will try to open a new connection (database session), which produces a concurrency problem if you already have one open (that `Engine` interface that's being passed around).

I think you'll be alright with this (it now passes the test in my computer).

**Comment by guillep2k** [comment]
Note: to apply that patch, simple unzip and use: `git apply 8086.diff`

**Comment by richmahn** [comment]
@guillep2k Thank! Applying and trying now.

**Comment by richmahn** [comment]
@guillep2k I tried the exact command you gave and I get this:

>  $ git apply 8086.diff
> error: patch failed: models/repo.go:323
> error: models/repo.go: patch does not apply
> error: patch failed: models/user.go:237
> error: models/user.go: patch does not apply

Any ideas? I try applying manually.

**Comment by guillep2k** [comment]
You can just check the changes from my repo:

**Comment by guillep2k** [comment]
You can just check the changes from my fork:
https://github.com/guillep2k/gitea/commit/8357de2c5431330ab562863375d8c6cca99763f0

**Comment by guillep2k** [comment]
I could have PR'd *your* repo, but for that I had to fork from it and I wasn't sure Github would let me have two giteas.

**Comment by guillep2k** [comment]
The fail you've just got should be fixed in a recent commit. Please use the update branch button below "Mergin is blocked" to check again.

File: go.mod [hunk 10]
```diff
@@ -75,6 +75,8 @@ require (
 	github.com/mattn/go-sqlite3 v1.11.0
 	github.com/mcuadros/go-version v0.0.0-20190308113854-92cdf37c5b75
 	github.com/microcosm-cc/bluemonday v0.0.0-20161012083705-f77f16ffc87a
+	github.com/modern-go/concurrent v0.0.0-20180306012644-bacd9c7ef1dd // indirect
```
Comment (lunny) [inline go.mod]: Please use go1.13 to compile.

File: modules/structs/user.go [hunk 11]
```diff
@@ -20,16 +20,52 @@ type User struct {
...
+	PubicRepos int64 `json:"public_repos"`
+	// Followers count
+	Followers int `json:"followers"`
+	// Following count
+	Following int `json:"following"`
 	// User locale
 	Language string `json:"language"`
 	// Is the user an administrator
-	IsAdmin bool `json:"is_admin"`
+	IsAdmin bool `json:"site_admin"`
```
[the remark this answers is not in the document]
Comment (richmahn) [reply modules/structs/user.go]: ah, you're right. Putting that back. Thanks for finding that.

File: go.mod (hunk 10 above)
Comment (richmahn) [reply go.mod]: I upgraded, but when I did a `make generate` and `make build` the go.sum file remained the same. So I just reverted back to master's go.sum and built again and works. Committed the reverted go.sum file. Was that the right thing to do @lunny?

Comment (guillep2k) [reply go.mod]: I think that happens when you update the tools as well (e.g. golanci-lint). Try building again and see if it changes again.

File: models/user.go [hunk 12]
```diff
@@ -226,16 +237,40 @@ func (u *User) GetEmail() string {
...
+		ID:               u.ID,
+		UserName:         u.Name,
+		FullName:         u.FullName,
+		Email:            u.GetEmail(),
+		HideEmail:        u.KeepEmailPrivate,
+		AvatarURL:        u.AvatarLink(),
+		URL:              apiURL,
+		HTMLURL:          setting.AppURL + u.LowerName,
+		FollowersURL:     apiURL + "/followers",
+		FollowingURL:     apiURL + "/followers",
```
Comment (6543) [inline models/user.go]: change followers -> following

**Review by 6543 — CHANGES_REQUESTED** [review]
@richmahn smal copy-paste issue i think

**Comment by 6543** [comment]
and at the general discustion: be more private than github is a pro from my perspective

File: models/user.go (hunk 12 above)
Comment (richmahn) [reply models/user.go]: Oh wow, thanks.

Comment (richmahn) [reply models/user.go]: Fixed

**Comment by 6543** [comment]
I'll test this ...

**Comment by 6543** [comment]
lint error :(

**Comment by 6543** [comment]
@richmahn 
```

# code.gitea.io/gitea/modules/notification/webhook
--
587 | modules/notification/webhook/webhook.go:91:4: cannot use u.APIFormat() (type *structs.User) as type *structs.Organization in field value
588 | modules/notification/webhook/webhook.go:107:4: cannot use u.APIFormat() (type *structs.User) as type *structs.Organization in field value
589 | modules/notification/webhook/webhook.go:124:4: cannot use u.APIFormat() (type *structs.User) as type *structs.Organization in field value
```

**Comment by richmahn** [comment]
@6543 Yes, seems like changes have been made to the same thing I'm working on. No problem. Just been busy this week on another project. Will fix that soon.

File: models/org_team.go [hunk 13]
```diff
@@ -37,6 +39,27 @@ type Team struct {
 	CanCreateOrgRepo        bool        `xorm:"NOT NULL DEFAULT false"`
 }
 
+// AfterLoad is invoked from XORM after setting the values of all fields of this object.
+func (t *Team) AfterLoad() {
```
Comment (lunny) [inline models/org_team.go:43]: It's better to use `AfterLoad(sess *xorm.Session)` and resue `see` to load organization.

File: modules/convert/convert.go [hunk 14]
```diff
@@ -229,52 +227,44 @@ func ToDeployKey(apiLink string, key *models.DeployKey) *api.DeployKey {
...
-		AvatarURL:                 org.AvatarLink(),
-		UserName:                  org.Name,
-		FullName:                  org.FullName,
-		Description:               org.Description,
-		Website:                   org.Website,
-		Location:                  org.Location,
-		Visibility:                org.Visibility.String(),
-		RepoAdminChangeTeamAccess: org.RepoAdminChangeTeamAccess,
-	}
+	return org.APIFormatOrganization()
```
Comment (lunny) [inline modules/convert/convert.go:230]: Our long purpose here is to move all convertion from package `models` to `modules/convert`.

File: modules/structs/user.go [hunk 15]
```diff
@@ -9,27 +9,95 @@ import (
...
+// User represents the brief API response fields for models.User
 // swagger:model
 type User struct {
 	// the user's id
 	ID int64 `json:"id"`
+	// the user's login
+	Login string `json:"login"`
 	// the user's username
-	UserName string `json:"login"`
+	UserName string `json:"username"`
```
Comment (lunny) [inline modules/structs/user.go:20]: What's the different from `username` and `login`?

**Review by lunny — CHANGES_REQUESTED** [review]

**Comment by lunny** [comment]
Please resolve the conflicts.

**Comment by richmahn** [comment]
This has gotten a far behind and probably should be reworked with the latest master.
